FABulous/fabric_generator/file_parser.py

```diff
@@ -664,6 +653,82 @@ def parseFile(filename: str, belPrefix: str = "", filetype: str = "") -> Bel:
         isConfig = False
         isShared = False
 
+    if filetype == "verilog":
+        # Runs yosys on verilog file, creates netlist, saves to json in same directory.
+        json_file = filename.replace(".v", ".json")
+        runCmd = [
+            "yosys",
+            "-qp"
+            f"read_verilog {filename}; proc -noopt; write_json -compat-int {json_file}",
+        ]
+        try:
+            subprocess.run(runCmd, check=True)
+        except subprocess.CalledProcessError as e:
+            logger.error(f"Failed to run yosys command: {e}")
+            raise ValueError
+
+        with open(f"{json_file}", "r") as f:
+            data_dict = json.load(f)
+        # Default yosys list names added.
+        modules = data_dict.get("modules", {})
+        sub_modules = ["parameter_default_values", "cells", "memories", "netnames"]
+        filtered_ports: Dict[str, IO] = {}
+        # Gatheres port name and direction, filters out configbits as they show in ports.
+        for module_name, module_info in modules.items():
+            ports = module_info["ports"]
+            for port_name, details in ports.items():
+                if "ConfigBits" in port_name:
+                    continue
+                if "UserCLK" in port_name:
+                    userClk = True
+                if port_name[-1].isdigit():
+                    individually_declared = True
+                direction = IO[details["direction"].upper()]
+                bits = details.get("bits", [])
+                filtered_ports[port_name] = (direction, bits)
+        # Passed attributes dont show in port list, checks for attributes in other lists.
+        for submodule in sub_modules:
```

Comment:
I might be wrong, but I notice all the attributes are on the `netnames`

Reply:
My understanding from yosys documentation for [write_json](https://yosyshq.readthedocs.io/projects/yosys/en/latest/cmd/write_json.html) was that `cells` and `memories` also contain attribute details. I'm not sure if ports and declarations always end up in `netnames` but I thought it would be best to be thorough in-case through optimisation yosys does put them else where.

Reply:
Yes, it is also better to check, but for example, if `cells` will never contain the FABulous related attributes, then we will never need to check them.